### PR TITLE
8273000: Remove WeakReference-based class initialisation barrier implementation

### PR DESCRIPTION
**Pull request by iwanowww**
Get rid of WeakReference-based logic in DirectMethodHandle::checkInitialized() and reimplement it with `Unsafe::ensureClassInitialized()`/`shouldBeInitialized()`. 

The key observation is that `Unsafe::ensureClassInitialized()` does not block the initializing thread. 

Also, removed `Unsafe::shouldBeInitialized()` in `DMH::shouldBeInitialized(MemberName)` to save on calling into the VM.
`Unsafe::ensureClassInitialized()` already has a fast-path check which checks whether the class is fully initialized or not.

Testing: tier1 - tier6

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273000](https://bugs.openjdk.java.net/browse/JDK-8273000): Remove WeakReference-based class initialisation barrier implementation

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to 64f2de83abe834cdbf2c6c3511274a73c3f64756
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 9fc921675840daa8c94a4a666eb6480d654a7428

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5258/head:pull/5258` \
`$ git checkout pull/5258`

Update a local copy of the PR: \
`$ git checkout pull/5258` \
`$ git pull https://git.openjdk.java.net/jdk pull/5258/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5258`

View PR using the GUI difftool: \
`$ git pr show -t 5258`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5258.diff">https://git.openjdk.java.net/jdk/pull/5258.diff</a>

</details>

### CODE REVIEW

**Comment by mlchung** [comment]
> May i suggest that we add some JavaDoc to `ensureClassInitialized` describing the two cases of the calling thread is the initialization thread or not.

This is a good suggestion that also applies to `Lookup::ensureInitialized`.   I created https://bugs.openjdk.java.net/browse/JDK-8273194 to improve the javadoc for `Lookup::ensureInitialized`.

**Review by mlchung — APPROVED** [review]

**Comment by iwanowww** [comment]
> May i suggest that we add some JavaDoc to ensureClassInitialized

Thanks, Paul. How does the latest version look?

File: src/java.base/share/classes/jdk/internal/misc/Unsafe.java [hunk 1]
```diff
@@ -1143,9 +1143,13 @@ public boolean shouldBeInitialized(Class<?> c) {
...
-     * Ensures the given class has been initialized. This is often
-     * needed in conjunction with obtaining the static field base of a
-     * class.
+     * Ensures the given class has been initialized (see JVMS-5.5 for details).
+     * This is often needed in conjunction with obtaining the static field base
+     * of a class.
+     *
+     * The call returns when either class {@code c} is fully initialized or
+     * class {@code c} is being initialized and the call is performed from
+     * the initializing thread.
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/misc/Unsafe.java:1152]:
```suggestion
     * The call returns when either class {@code c} is fully initialized or
     * class {@code c} is being initialized and the call is performed from
     * the initializing thread. In the latter case a subsequent call to
     * {@link #shouldBeInitialized}, from the calling thread of this call,
     * will return {@code true}.
```

**Review by PaulSandoz — APPROVED** [review]
Looks good, just a minor suggestion up to you to accept or not.

File: src/java.base/share/classes/jdk/internal/misc/Unsafe.java (hunk 1 above)
Comment (dholmes-ora) [reply src/java.base/share/classes/jdk/internal/misc/Unsafe.java:1152]: Aren't "the calling thread of this call" and "the initializing thread" the same thread in the latter case?

Comment (iwanowww) [reply src/java.base/share/classes/jdk/internal/misc/Unsafe.java]: Agree. I dropped "from the calling thread of this call" part and incorporated the rest into the latest version.

**Comment by iwanowww** [comment]
Thanks for the reviews, Mandy, Paul, and David.

/integrate

**Pull request closed without merging**